### PR TITLE
nixos/network: add gre virtual interfaces

### PR DESCRIPTION
**Pull request by hexagonal-sun**
###### Motivation for this change
Add options that allow `gre` tunnels to be creating in the nixOS config.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/tasks/network-interfaces-scripted.nix [hunk 1]
```diff
@@ -253,7 +253,11 @@ let
             '';
           };
 
-        createTunDevice = i: nameValuePair "${i.name}-netdev"
+          createTunDevice = i: nameValuePair "${i.name}-netdev"
```
Comment (pennae) [inline nixos/modules/tasks/network-interfaces-scripted.nix:256]:
```suggestion
        createTunDevice = i: nameValuePair "${i.name}-netdev"
```
did your editor indent this by mistake?

File: nixos/modules/tasks/network-interfaces-scripted.nix [hunk 2]
```diff
@@ -253,7 +253,11 @@ let
...
+          ( let
+              greType = if i.virtualType == "gretun" then "gre" else i.virtualType;
+              greLocal = optionalString (i.gre.local != null) " local ${i.gre.local}";
```
Comment (pennae) [inline nixos/modules/tasks/network-interfaces-scripted.nix:259]: this entire block could be moved to the `script` attribute that uses it. would probably be easier to read like that, adding lets like this is rather uncommon

**Review by pennae — CHANGES_REQUESTED** [review]
this also needs to be added to the networkd backend and tested for both backends

File: nixos/modules/tasks/network-interfaces-scripted.nix (hunk 2 above)
Comment (hexagonal-sun) [reply nixos/modules/tasks/network-interfaces-scripted.nix:259]: Yep, that clears things up. Done.

File: nixos/modules/tasks/network-interfaces-scripted.nix (hunk 1 above)
Comment (hexagonal-sun) [reply nixos/modules/tasks/network-interfaces-scripted.nix:256]: Whops. Yeah my editor added the indent.  Removed.

File: nixos/modules/tasks/network-interfaces.nix [hunk 3]
```diff
@@ -1215,6 +1230,12 @@ in
         message = ''
           Setting a MAC Address for tun device ${i.name} isn't supported.
         '';
+      })) ++ (forEach interfaces (i: {
+        assertion = (i.virtual && (i.virtualType == "gretun" ||
+                                   i.virtualType == "gretap")) -> i.gre.remote != null;
```
[the remark this answers is not in the document]
Comment (hexagonal-sun) [reply nixos/modules/tasks/network-interfaces.nix:1235]: Agreed.  Changed `gre.remote` to be `nullOr str`.

**Comment by hexagonal-sun** [comment]
> this also needs to be added to the networkd backend and tested for both backends

Added systemd back-end and tested with both.  Some small changes I've made atop of your comments:

 - I've renamed `gretun` to `gre`.  This is consistent with how other tools call this mode and it claens up the code from translating `gretun` -> `gre`.
 - I'm using `hasPrefix "gre"` to check whether the virtual type is GRE.

**Comment by pennae** [comment]
> Added systemd back-end and tested with both.

can you add a nixos test for the testing you've done as well? would be nice to ensure this keeps working even when other things are changed. apart from that, looks good to me :+1:

**Comment by hexagonal-sun** [comment]
@pennae Thanks for taking a look but, sorry, I decided to do a bit of a rewrite and ask for another review.  `gre` interfaces are more like `sit` interfaces, i.e. they're a point-to-point link which don't have an IP address, netmask, etc.  The previous configuration would allow the user to set an IP address on the GRE tunnel which doesn't make sense.  The new rewrite should fix this.  I'll also look at creating some tests.

Thanks!

File: nixos/modules/tasks/network-interfaces.nix [hunk 4]
```diff
@@ -996,6 +997,64 @@ in
...
+          greBridge = {
+            remote = "10.0.0.1";
+            local = "10.0.0.22";
+            dev = "enp4s0f0";
+            type = "tap";
+          };
+        }
+      '';
+      description = ''
+        This option allows you to define a Generic Routing Encapsulation (GRE) tunnel.
```
Comment (pennae) [inline nixos/modules/tasks/network-interfaces.nix:1013]:
```suggestion
        This option allows you to define Generic Routing Encapsulation (GRE) tunnels.
```

File: nixos/modules/tasks/network-interfaces.nix [hunk 5]
```diff
@@ -996,6 +997,64 @@ in
       });
     };
 
+    networking.gre = mkOption {
```
Comment (pennae) [inline nixos/modules/tasks/network-interfaces.nix:1000]: in keeping with other options this should be pluralized, but `gres` seems more confusing than helpful at first glance. how about
```suggestion
    networking.greTunnels = mkOption {
```

File: nixos/modules/tasks/network-interfaces-scripted.nix [hunk 6]
```diff
@@ -532,6 +532,34 @@ let
             '';
           });
 
+        createGreDevice = n: v: nameValuePair "${n}-netdev"
+          (let
+            deps = deviceDependency v.dev;
+            greType = if v.type == "tun" then "gre" else "gretap";
```
Comment (pennae) [inline nixos/modules/tasks/network-interfaces-scripted.nix:538]: this expression is duplicated a few times. you could add it to the option definition to have the option checked against the declared type, but stored for later use as though this expression had always been applied to it first.
```nix
  apply = v: if v == "tun" then "gre" else "gretap";
```

Comment (hexagonal-sun) [reply nixos/modules/tasks/network-interfaces-scripted.nix:538]: Nice!  I didn't know about this option.

**Comment by hexagonal-sun** [comment]
> > Added systemd back-end and tested with both.
> 
> can you add a nixos test for the testing you've done as well? would be nice to ensure this keeps working even when other things are changed. apart from that, looks good to me 👍

I've now added a gre test case which uses a GRE tunnel to get layer2 packets across VLANs.  Hope that's okay?

File: nixos/tests/networking.nix [hunk 7]
```diff
@@ -101,9 +101,6 @@ let
             { address = "192.168.1.3"; prefixLength = 32; }
             { address = "192.168.1.10"; prefixLength = 32; }
           ];
-          interfaces.eth2.ipv4.addresses = mkOverride 0 [
-            { address = "192.168.2.2"; prefixLength = 24; }
-          ];
```
Comment (pennae) [inline nixos/tests/networking.nix:106]: is this an intentional change? removing this breaks the test.

**Review by pennae — COMMENTED** [review]
looking good! the test runs with both scripted and networkd setups, but looks like an unrelated change broke the static test. other than that, seems good to me.

File: nixos/tests/networking.nix (hunk 7 above)
Comment (hexagonal-sun) [reply nixos/tests/networking.nix]: Oops. Sorry that must have been while I was figuring out how the tests worked!  Should be reinstated now.

**Comment by hexagonal-sun** [comment]
> looking good! the test runs with both scripted and networkd setups, but looks like an unrelated change broke the static test. other than that, seems good to me.

Great!  The `static` test should now be fixed.

**Review by pennae — APPROVED** [review]

**Comment by hexagonal-sun** [comment]
@tomberek Are you okay to review please? If you're happy, are okay to merge this? Thanks! :-)

**Comment by hexagonal-sun** [comment]
@SuperSandro2000 If you're happy with the above, is it okay to get merged?  Thanks!

**Comment by symphorien** [comment]
@GrahamcOfBorg test networking

**Pull request merged**